File: python/feature_extract.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jan 25 16:25:14 2016

Perform feature extraction, veoctorized

@author: R. Armiger
"""

import numpy as np

def feature_extract(y, zc_thresh = 0.15, ssc_thresh = 0.15):
    #compute features
    #
    # Input: 
    # data buffer to compute features
    # y = numpy.zeros((numSamples, numChannels))
    # E.g. numpy.zeros((50, 8))
    #
    # Optional:
    # Thresholds for computing zero crossing and slope sign change features
    #
    # Output: feature vector should be [1,nChan*nFeat]
    # data ordering is as follows
    # [ch1f1, ch1f2, ch1f3, ch1f4, ch2f1, ch2f2, ch2f3, ch2f4, ... chNf4]

    
    

    # Number of Samples
    n = y.shape[0]

    # Normalize features so they are independant of the window size    
    Fs=200

    # Value to compute 'zero-crossing' around
    t=0.0
        
    # Compute MAV across all samples (axis=0)
    MAV = np.mean(abs(y),0)     # MAV shouldn't be normalized
    
    # Curve length is the sum of the absolute value of the derivative of the 
    # signal, normalized by the sample rate
    LEN = np.sum(abs( np.diff(y, axis=0) ), axis=0) * Fs / n

    # Criteria for crossing zero
    # zeroCross=(y[iSample] - t > 0 and y[iSample + 1] - t < 0) or (y[iSample] - t < 0 and y[iSample + 1] - t > 0)
    # overThreshold=abs(y[iSample] - t - y[iSample + 1] - t) > zc_thresh
    # if zeroCross and overThreshold:
    #     # Count a zero cross
    #     ZC[iChannel]=ZC[iChannel] + 1
    ZC = np.sum(\
         ((y[0:n-1,:] - t > 0) & (y[1:n,:] - t < 0) | \
         (y[0:n-1,:] - t < 0) & (y[1:n,:] - t > 0)) & \
         (abs(y[0:n-1,:] - t - y[1:n,:] - t) > zc_thresh)
         ,axis=0) * Fs / n

    # Criteria for counting slope sign changes
    # signChange = (y[iSample] > y[iSample - 1]) and (y[iSample] > y[iSample + 1]) or (y[iSample] < y[iSample - 1]) and (y[iSample] < y[iSample + 1])
    # overThreshold=abs(y[iSample] - y[iSample + 1]) > ssc_thresh or abs(y[iSample] - y[iSample - 1]) > ssc_thresh
    # if signChange and overThreshold:
    #     # Count a slope change
    #     SSC[iChannel]=SSC[iChannel] + 1
    SSC = np.sum(\
          ((y[1:n-1,:] > y[0:n-2,:]) & (y[1:n-1,:] > y[2:n,:]) | \
          (y[1:n-1,:] < y[0:n-2,:]) & (y[1:n-1,:] < y[2:n,:])) & \
          ((abs(y[1:n-1,:] - y[2:n,:]) > ssc_thresh) | (abs(y[1:n-1,:] - y[0:n-2,:]) > ssc_thresh)) \
          ,axis=0) * Fs / n

         
    #VAR = np.var(y,axis=0) * Fs / n
    
    features=np.vstack((MAV,LEN,ZC,SSC))
    
    return features.T.reshape(1, 32)

# Offline test code
#
#
#import matplotlib.pyplot as plt
#import math
#import timeit
#
#NUM = 2000
#emg_buffer = np.zeros((NUM,8))
#sinArray = np.sin(2*math.pi*10*np.linspace(0,1,num=NUM))
#
#emg_buffer[:,:1] = np.reshape(sinArray,(NUM,1))
#emg_buffer[:,:7] = np.reshape(sinArray,(NUM,1))
#plt.plot(sinArray)
#
#
#start_time = timeit.default_timer()
## code you want to evaluate
#f = feature_extract(emg_buffer)
## code you want to evaluate
#elapsed = timeit.default_timer() - start_time
#print(elapsed)
#
#print(f)

File: Utilities/feature_extract.m
function features = feature_extract(windowData,windowSize,zc_thresh,ssc_thresh) %% codegen
% Extract time domain feature extraction function
% Note that this is not an exact copy of the previous methodology,
% but RSA tried and tested this code with minimal differences
% Inputs:
%   windowData - Buffer of windowed EMG data, typically 16x150
% Outputs:
%   features - [numChannels 4] matrix of features;
%       order: [MAV(:) LEN(:) ZC(:) SSC(:)]
%
% R. Armiger 07-July-2009: Revised to isolate feature extraction functionality only
%                  Signal magnitude will be extracted on a per classifier
%                  basis.  Also set to inherit class type and made
%                  windowSize tunable
% R. Armiger 30-Nov-2009: Updated to include thresholding on zero-crossing
%       and slope sign changes
% R. Armiger 09-Feb-2016: Vectorized code to improve speed:
%       Before: Elapsed time is 0.001198 seconds.
%       After:  Elapsed time is 0.000358 seconds.
%

%  Thresholds for computing zero crossing and slope sign change features

if nargin < 4
    ssc_thresh = 0.15;
end
if nargin < 3
    zc_thresh = 0.15;
end

[nChannels, nSamples] = size(windowData);

windowSize = max(windowSize,10);
windowSize = min(windowSize,nSamples);

% VAR = zeros(nChannels,1,myType);    % variance

%  Value to compute 'zero-crossing' around
t = 0.0;

%  Normalize features so they are independant of the window size
Fs = 1000;
n = windowSize;

idStart = (1+nSamples-windowSize);
y = windowData(1:nChannels,idStart:end)';

%  MAV shouldn't be normalized
MAV = mean( abs( y ) );

%  Curve length is the sum of the absolute value of the derivative of the
%  signal, normalized by the sample rate

LEN = sum( abs( diff(y) ) ) * Fs / n;


%  Criteria for crossing zero
%  zeroCross=(y[iSample] - t > 0 and y[iSample + 1] - t < 0) or (y[iSample] - t < 0 and y[iSample + 1] - t > 0)
%  overThreshold=abs(y[iSample] - t - y[iSample + 1] - t) > zc_thresh
%  if zeroCross and overThreshold:
%      %  Count a zero cross
%      ZC[iChannel]=ZC[iChannel] + 1

ZC = sum(...
    ((y(1:n-1,:) - t > 0) & (y(2:n,:) - t < 0) | ...
    (y(1:n-1,:) - t < 0) & (y(2:n,:) - t > 0)) & ...
    (abs(y(1:n-1,:) - t - y(2:n,:) - t) > zc_thresh) ...
    ) * Fs / n;


%  Criteria for counting slope sign changes
%  signChange = (y[iSample] > y[iSample - 1]) and (y[iSample] > y[iSample + 1]) or (y[iSample] < y[iSample - 1]) and (y[iSample] < y[iSample + 1])
%  overThreshold=abs(y[iSample] - y[iSample + 1]) > ssc_thresh or abs(y[iSample] - y[iSample - 1]) > ssc_thresh
%  if signChange and overThreshold:
%      %  Count a slope change
%      SSC[iChannel]=SSC[iChannel] + 1

SSC = sum( ...
    ((y(2:n-1,:) > y(1:n-2,:)) & (y(2:n-1,:) > y(3:n,:)) |  ...
    (y(2:n-1,:) < y(1:n-2,:)) & (y(2:n-1,:) < y(3:n,:))) &  ...
    ((abs(y(2:n-1,:) - y(3:n,:)) > ssc_thresh) | (abs(y(2:n-1,:) - y(1:n-2,:)) > ssc_thresh))...
    ) * Fs / n;

features = [MAV(:) LEN(:) ZC(:) SSC(:)];
